chunk.go

```diff
@@ -5,41 +5,426 @@ import (
+	// n-byte buffer here, but while we wait to send that we gain
+	// more capacity in our write channel. In such cases it would
+	// be preferable for us to to increase the amount of data that
+	// we include in this chunk.
+	//
+	// We can leave this as an optimisation for later though
+	return f.sendChunk(writerChunk{
+		buf:     buf,
+		advance: nb,
+	})
+}
+
+// satisfies io.Writer
+func (f *flusherWriter) Write(buf []byte) (int, error) {
+	if f.full == nil {
+		return 0, io.ErrClosedPipe
+	}
+
+	// Pass large buffers directly to the draining side
+	if len(buf) > len(f.buf) {
+		if err := f.sendChunk(writerChunk{buf: buf}); err != nil {
+			return 0, err
+		}
+
+		if err := f.waitTilProcessed(buf); err != nil {
+			return 0, err
+		}
+		return len(buf), nil
+	}
+
+	// Handle shorter writes by copying through our internal buffer
+	totalWritten := 0
+	for len(buf) > 0 {
+		destBuf, err := f.getWriteBuffer()
+		if err != nil {
+			return totalWritten, err
+		}
+
+		nb := copy(destBuf, buf)
+		if err := f.written(nb); err != nil {
+			return totalWritten, err
+		}
+		totalWritten += nb
+	}
+	return totalWritten, nil
+}
+
+// satisfies http.Flusher
+func (f *flusherWriter) Flush() {
+	// Check if we're closed
+	if f.full == nil {
+		return
+	}
+	// Since we don't have an error return here, we have to discard
+	// any returned error. That's OK though: a future Write or Close
+	// will find it
+	_ = f.sendChunk(writerChunk{buf: nil, advance: 0})
+}
+
+// satisifes io.ReaderFrom
+func (f *flusherWriter) ReadFrom(r io.Reader) (n int64, err error) {
+	// Check if we are closed
+	if f.full == nil {
+		return 0, io.ErrClosedPipe
+	}
+
+	for {
+		var destBuf []byte
+		var nb int
+
+		destBuf, err = f.getWriteBuffer()
+		if err != nil {
 			break
 		}
+
+		nb, err = r.Read(destBuf)
+		n += int64(nb)
+		if nb == 0 || err != nil {
+			break
+		}
+
+		err = f.written(nb)
+		if err != nil {
+			break
+		}
+	}
+
+	if err == io.EOF {
+		err = nil
+	}
+
+	return n, err
+}
+
+// satisfies io.WriteCloser
+// Closure happens from the source side, and then waits for
+// the destination side to drain
+func (f *flusherWriter) Close() error {
+	if f.full != nil {
+		close(f.full)
+		for _ = range f.empty {
+			// Pump the returns channel until closure
+		}
+
+		// For inexplicable reasons, reading from a nil channel blocks
+		// forever. Wait until run() has finished before clearing this
+		// (our closure signal)
+		f.full = nil
+	}
+	// Return any stored error
+	return f.err
+}
+
+//
+// Destination side logic
+//
+
+func (f *flusherWriter) flushDestination() {
+	if flusher, ok := f.dst.(http.Flusher); ok {
+		flusher.Flush()
+	}
+}
+
+// writeChunk handles writing a chunk to our destination stream
+func (f *flusherWriter) writeChunk(chunk writerChunk) (flushed bool, err error) {
+	// If there's no buf, this is a flush
+	if chunk.buf == nil {
+		f.flushDestination()
+		return true, nil
+	}
+
+	// Otherwise, this is a write
+	nw, err := f.dst.Write(chunk.buf)
```

Comment:
We're calling out to the streamer here in a background goroutine, so any panic() inside will crash us.

This probably isn't a concern (we trust net/http, I'd guess?) but a little thing to be aware of. Ideally we might wrap this in some sort of teardown/recover wrapper